File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      silence?(input) ->
        "Fine. Be that way!"

      yell?(input) ->
        if question?(input), do: "Calm down, I know what I'm doing!", else: "Whoa, chill out!"

      question?(input) ->
        "Sure."

      true ->
        "Whatever."
    end
  end

  defp question?(input), do: String.ends_with?(input, "?")
  defp yell?(input), do: String.upcase(input) == input && (String.downcase(input) != input)
  defp silence?(input), do: String.trim(input) == ""
end

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.String.Utils (strip, endswith)
import Data.Char

responseFor :: String -> String
responseFor xs
  | isSilence strippedInput = "Fine. Be that way!"
  | isYell strippedInput && isQuestion strippedInput = "Calm down, I know what I'm doing!"
  | isYell strippedInput = "Whoa, chill out!"
  | isQuestion strippedInput = "Sure."
  | otherwise = "Whatever."
  where
    strippedInput = strip xs

isSilence :: String -> Bool
isSilence = null

isYell :: String -> Bool
isYell xs = all(\x -> toUpper x == x) xs && any(\x -> toLower x /= x) xs

isQuestion :: String -> Bool
isQuestion = endswith "?"
